plugins/editorconfig/src/org/editorconfig/language/services/impl/EditorConfigOptionDescriptorStorage.kt

```diff
@@ -0,0 +1,45 @@
+// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package org.editorconfig.language.services.impl
+
+import org.editorconfig.language.schema.descriptors.EditorConfigDescriptor
+import org.editorconfig.language.schema.descriptors.impl.EditorConfigConstantDescriptor
+import org.editorconfig.language.schema.descriptors.impl.EditorConfigOptionDescriptor
+import org.editorconfig.language.schema.descriptors.impl.EditorConfigQualifiedKeyDescriptor
+import org.editorconfig.language.schema.descriptors.impl.EditorConfigUnionDescriptor
+import org.editorconfig.language.util.ignore
+
+class EditorConfigOptionDescriptorStorage(source: Iterable<EditorConfigOptionDescriptor>) {
+  private val simpleKeyDescriptors: Map<String, EditorConfigOptionDescriptor>
+  private val complexKeyDescriptors: List<EditorConfigOptionDescriptor>
+
+  init {
+    val simpleKeyDescriptors = EditorConfigIgnoreCaseMap<EditorConfigOptionDescriptor>()
+    val complexKeyDescriptors = mutableListOf<EditorConfigOptionDescriptor>()
+
+    fun insert(descriptor: EditorConfigOptionDescriptor, key: EditorConfigDescriptor): Unit = when (key) {
+      is EditorConfigConstantDescriptor ->
+        simpleKeyDescriptors[key.text] = simpleKeyDescriptors[key.text]?.merge(descriptor) ?: descriptor
+      is EditorConfigUnionDescriptor -> key.children.forEach { insert(descriptor, it) }
+      is EditorConfigQualifiedKeyDescriptor -> complexKeyDescriptors.add(descriptor).ignore
+      else -> Unit
+    }
+
+    source.forEach { insert(it, it.key) }
+    this.simpleKeyDescriptors = simpleKeyDescriptors
+    this.complexKeyDescriptors = complexKeyDescriptors
+  }
+
+  val optionDescriptors by lazy { simpleKeyDescriptors.values + complexKeyDescriptors }
+
+  operator fun get(key: String) = simpleKeyDescriptors[key]
+  operator fun get(key: List<String>) =
```

Comment:
Don't we need to get more optimal here, with like sort of trie or something? how many complex key are there for roslyn, and how often are they retrieved?